### PR TITLE
[MRG] Add UNHaP experiments and documentation

### PR DESCRIPTION
**Pull request by vloison**
To do : 
[x] homogenize docstring of FaDIn and UNHaP
[x] Add solvers benchmarked in the  UNHaP paper
[x] Add UNHaP experiments
[x] Update README
[ ] Update pyproject.toml with optional dependencies for experiments.
[ ] More beginner-friendly tutorials.
[ ] More doc for UNHaP and its utility functions.

To do in a future PR:
[ ] Factorize UNHaP and FaDIn utility functions
This will not be so easy. For instance, during optimization, model parameters are stored as a list. And UNHaP has one more model parameter, the noise baseline. Another example: for fadin, events are floats, for unhap, they are arrays of size two. This means that between the two solvers, several functions have lots of code in common but are tricky to factorize robustly.
[x] add tests for UNHaP and FaDIn:
-  test masks of UNHaP and FaDIn
- test random init and moment matching of UNHaP and FaDIn

### CODE REVIEW

**Comment by vloison** [comment]
@tomMoral, now would be a good time for a review, before the number of lines changes is too big :)

The experiment files `run_ecg` and `utils_ecg` don't need reviewing, in my opinion.

File: README.md [hunk 1]
```diff
@@ -1,17 +1,29 @@
 # FaDIn: Fast Discretized Inference For Hawkes Processes with General Parametric Kernels
+# UNHaP: Unmixing Noise from Hawkes Processes
```
Comment (tomMoral) [inline README.md:2]:
```suggestion
# FaDIn: a tool box for fast and robust inference for parametric point processes
```

Comment (tomMoral) [reply README.md:2]: Then, the papers are cited below

File: README.md [hunk 2]
```diff
@@ -41,6 +53,12 @@ pip install -e ".[dev]"
 pre-commit install
 ```
 
+## Short examples
+A few illustrative examples are provided in the `examples` folder of this repository:
```
Comment (tomMoral) [inline README.md:57]:
```suggestion
A few illustrative examples are provided in the `examples` folder of this repository, in particular:
```

File: fadin/init.py [hunk 3]
```diff
@@ -240,6 +240,7 @@ def init_hawkes_params_unhap(solver, init, events, n_ground_events,
         'alpha' and 'kernel'.
     events: list of array of size number of timestamps,
         list size is self.n_dim.
+    events_grid: TODO
```
Comment (tomMoral) [inline fadin/init.py:243]: trailling todo

File: fadin/loss_and_gradient.py [hunk 4]
```diff
@@ -35,14 +35,17 @@ def compute_gradient_fadin(solver, events_grid, discretization,
...
+        discrete_l2_loss_precomputation(
+            solver.zG,
+            solver.zN,
+            solver.ztzG,
+            solver.params_intens[0],
+            solver.params_intens[1],
+            kernel, n_events,
+            solver.delta,
+            end_time
+            ).detach()
```
Comment (tomMoral) [inline fadin/loss_and_gradient.py:48]:
```suggestion
    solver.v_loss[i] = discrete_l2_loss_precomputation(
        solver.zG,
        solver.zN,
        solver.ztzG,
        solver.params_intens[0],
        solver.params_intens[1],
        kernel, n_events,
        solver.delta,
        end_time
    ).detach()
```

File: fadin/solver.py [hunk 5]
```diff
@@ -113,21 +116,39 @@ class FaDIn(object):
...
 
+    random_state : `int`, `RandomState` instance or `None`, `default=None`
+        Set the torch seed to 'random_state'.
+        If set to `None`, torch seed will be set to 0.
+
     Attributes
     ----------
-    param_baseline : `tensor`, shape (n_dim)
-        Baseline parameter of the Hawkes process.
+    params_intens : `list` of `tensor`
```
Comment (tomMoral) [inline fadin/solver.py:125]: Attributes that are only available after fit should finish with `_`:
```suggestion
    params_intens_ : `list` of `tensor`
```

File: fadin/solver.py [hunk 6]
```diff
@@ -690,8 +718,17 @@ def fit(self, events, end_time):
                 precomputations = compute_constants_unhap(
                     z_tilde, marks_grid, events_grid,
                     self.param_rho, marked_quantities[1], self.L)
-
-                self.param_rho = torch.round(self.params_mixture[0].detach())
+                if self.stoc_classif is False:
+                    # vanilla UNHaP
+                    self.param_rho = torch.round(self.params_mixture[0].detach())
+                if self.stoc_classif is True:
```
Comment (tomMoral) [inline fadin/solver.py:724]:
```suggestion
                if not self.stoc_classif:
                    # vanilla UNHaP
                    self.param_rho = torch.round(self.params_mixture[0].detach())
                else:
```

File: fadin/tests/test_init.py [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    ]
+    labels[0][-marked_events[0].shape[0]:] = 1.0
+    true_rho = [labels[i][events[i][:, 0].argsort()] for i in range(n_dim)]
+
+    return events_cat, noisy_marks, true_rho
+
+
+ev, noisy_marks, true_rho = simulate_marked_data(
+    baseline, baseline_noise.item(), alpha, end_time, seed=0
+)
```
Comment (tomMoral) [inline fadin/tests/test_init.py:89]: Why define it outside the function?
This is usually bad practice as this make the test collection longer than necessary if you filter out this test.

**Review by tomMoral — APPROVED** [review]
A few comments but overall LGTM

File: fadin/tests/test_init.py (hunk 7 above)
Comment (vloison) [reply fadin/tests/test_init.py:89]: It's a copy/paste of a function defined in `examples/plot_unhap`, I was thinking of adding it ti `utils/utils_simu`, what do you think?

Comment (tomMoral) [reply fadin/tests/test_init.py:89]: Yes you can put the function in an `utils`, that's a good idea.
But my comment was related to: the line 87-89, which instanciate `ev, noisy_marks, true_rho`, should be in the test, not in the module definition.

File: fadin/solver.py [hunk 8]
```diff
@@ -369,67 +395,79 @@ class UNHaP(object):
...
+        loss accross iterations.
+        If no early stopping, `n_iter` is equal to `max_iter`.
+
     """
     def __init__(self, n_dim, kernel, init='random', optim_mask=None,
                  kernel_length=1, delta=0.01, optim='RMSprop',
                  params_optim=dict(), max_iter=2000, batch_rho=100,
                  ztzG_approx=True, tol=10e-5, density_hawkes='linear',
-                 density_noise='uniform', moment_matching=False,
+                 density_noise='uniform', stoc_classif=False,
```
[the remark this answers is not in the document]
Comment (vloison) [reply fadin/solver.py:485]: nice catch, I added it

File: fadin/solver.py (hunk 5 above)
Comment (vloison) [reply fadin/solver.py:125]: okay thanks, will do ! these shoudl only be accessed when `fit` was called, should I add a `fitted` private attribute and an assert in the definition of `alpha`?

Comment (tomMoral) [reply fadin/solver.py:125]: you can yes :)

File: fadin/utils/utils_simu.py [hunk 9]
```diff
@@ -227,15 +241,18 @@ def simu_multi_poisson(end_time, intensity, upper_bound=None, random_state=None)
...
+    """ Simulate a multivariate Hawkes process following an immigration-birth
+    procedure.
         Edge effects may be reduced according to the second references below.
 
     References:
 
-    Møller, J., & Rasmussen, J. G. (2006). Approximate simulation of Hawkes processes.
+    Møller, J., & Rasmussen, J. G. (2006). Approximate simulation of Hawkes
+    processes.
     Methodology and Computing in Applied Probability, 8, 53-64.
```
Comment (tomMoral) [inline fadin/utils/utils_simu.py:252]:
```suggestion
    Møller, J., & Rasmussen, J. G. (2006). Approximate simulation of Hawkes
    processes. Methodology and Computing in Applied Probability, 8, 53-64.
```

File: fadin/utils/utils_simu.py [hunk 10]
```diff
@@ -227,15 +241,18 @@ def simu_multi_poisson(end_time, intensity, upper_bound=None, random_state=None)
...
 
-    Møller, J., & Rasmussen, J. G. (2005). Perfect simulation of Hawkes processes.
+    Møller, J., & Rasmussen, J. G. (2005). Perfect simulation of Hawkes
+    processes.
     Advances in applied probability, 37(3), 629-646.
```
Comment (tomMoral) [inline fadin/utils/utils_simu.py:256]:
```suggestion
    Møller, J., & Rasmussen, J. G. (2005). Perfect simulation of Hawkes
    processes. Advances in applied probability, 37(3), 629-646.
```

File: fadin/utils/utils_simu.py [hunk 11]
```diff
@@ -167,10 +178,13 @@ def simu_poisson(end_time, intensity, upper_bound=None, random_state=None):
...
+def simu_multi_poisson(end_time, intensity, upper_bound=None,
+                       random_state=None):
     """Simulate multivariate Poisson processes on [0, end_time] with
-    the Ogata's modified thinning algorithm by superposition of univariate processes.
-    If the intensity is a numerical value, simulate a Homegenous Poisson Process,
+    the Ogata's modified thinning algorithm by superposition of univariate
+    processes.
+    If the intensity is a numerical value, simulate a Homegenous Poisson
+    Process.
     If the intensity is a function, simulate an Inhomogenous Poisson Process.
```
Comment (tomMoral) [inline fadin/utils/utils_simu.py:188]:
```suggestion
    the Ogata's modified thinning algorithm by superposition of univariate
    processes. If the intensity is a numerical value, simulate a Homegenous
    Poisson Process. If the intensity is a function, simulate an
    Inhomogenous Poisson Process.
```

File: fadin/utils/utils_simu.py [hunk 12]
```diff
@@ -227,15 +241,18 @@ def simu_multi_poisson(end_time, intensity, upper_bound=None, random_state=None)
 def simu_hawkes_cluster(end_time, baseline, alpha, kernel,
                         params_kernel=dict(), kernel_length=None,
                         upper_bound=None, random_state=None):
-    """ Simulate a multivariate Hawkes process following an immigration-birth procedure.
+    """ Simulate a multivariate Hawkes process following an immigration-birth
+    procedure.
         Edge effects may be reduced according to the second references below.
```
Comment (tomMoral) [inline fadin/utils/utils_simu.py:246]:
```suggestion
    """ Simulate a multivariate Hawkes process following an immigration-birth
    procedure. Edge effects may be reduced according to the second
    references below.
```

**Review by tomMoral — COMMENTED** [review]
A few comment but otherwise, lgtm to merge like this

File: pyproject.toml [hunk 13]
```diff
@@ -41,6 +41,9 @@ test = [
     "pytest>=7.2",
 ]
 
+experiments = [
+    "dicodile>=0.3",
```
[the remark this answers is not in the document]
Comment (vloison) [reply pyproject.toml:45]: Yes exactly. There are a few packages we need just for the ECG and gait experiments. I'll add them here.

**Comment by vloison** [comment]
This PR is now ready to merge on my side, @tomMoral do you want to check the latest updates or is it good to go? :)
